Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp

```diff
@@ -19,96 +19,373 @@
+	if (!bSuccefullyCreatedReducedGrid)
+	{
+		return;
+	}
+
+	// Sort the spawn points by "distance from centre of cluster" ascending.
+	// This ensures we iterate from closest to furthest spawn point.
+	SpawnPoints.Sort([this](const FVector& Point1, const FVector& Point2)
+	{
+		const float Distance1 = FVector::Distance(Point1, WorldPosition);
+		const float Distance2 = FVector::Distance(Point2, WorldPosition);
+		return Distance2 > Distance1;
+	});
+
+	//Remove any excess spawn points
+	SpawnPoints.RemoveAt(MaxSpawnPoints, SpawnPoints.Num() - MaxSpawnPoints);
 }
 
-void ABenchmarkGymGameMode::GenerateTestScenarioLocations()
+const TArray<AActor*>& USpawnCluster::CreateSpawnPointActors()
 {
-	constexpr float RoamRadius = 7500.0f; // Set to half the NetCullDistance
+	UWorld* World = GetWorld();
+
+	for (int i = 0; i < SpawnPoints.Num(); ++i)
 	{
-		FRandomStream PlayerStream;
-		PlayerStream.Initialize(FCrc::MemCrc32(&ExpectedPlayers, sizeof(ExpectedPlayers))); // Ensure we can do deterministic runs
-		for (int i = 0; i < ExpectedPlayers; i++)
-		{
-			FVector PointA = PlayerStream.VRand()*RoamRadius;
-			FVector PointB = PlayerStream.VRand()*RoamRadius;
-			PointA.Z = PointB.Z = 0.0f;
-			PlayerRunPoints.Emplace(FBlackboardValues{ PointA, PointB });
-		}
+		const FVector& SpawnPoint = SpawnPoints[i];
+
+		FActorSpawnParameters SpawnInfo;
+		SpawnInfo.Owner = World->GetAuthGameMode();
+		SpawnInfo.bDeferConstruction = false;
+		SpawnInfo.SpawnCollisionHandlingOverride = ESpawnActorCollisionHandlingMethod::AlwaysSpawn;
+
+		// Swapping X and Y as GridBaseLBStrategy has the two reversed.
+		// Spawn actor is placed 3m off the ground to avoid spawning collisions.
+		FVector ModifiedSpawnPoint = SpawnPoint;
+		ModifiedSpawnPoint.X = SpawnPoint.Y;
+		ModifiedSpawnPoint.Y = SpawnPoint.X;
+		ModifiedSpawnPoint.Z = 300.0f;
+
+		UE_LOG(LogBenchmarkGymGameMode, Log, TEXT("Creating a new PlayerStart at location %s."), *ModifiedSpawnPoint.ToString());
+		SpawnPointActors.Add(World->SpawnActor<APlayerStart>(APlayerStart::StaticClass(), ModifiedSpawnPoint, FRotator::ZeroRotator, SpawnInfo));
 	}
+
+	return SpawnPointActors;
+}
+
+// --- USpawnArea ---
+
+void USpawnArea::GenerateSpawnClusters()
+{
+	TArray<FVector> ClusterPoints;
+	const bool bSuccefullyCreatedReducedGrid = GenerateMinimalGridInArea(Width, Height, MaxClusters, MinDistanceBetweenClusters, WorldPosition, ClusterPoints);
+	if (!bSuccefullyCreatedReducedGrid)
 	{
-		FRandomStream NPCStream;
-		NPCStream.Initialize(FCrc::MemCrc32(&TotalNPCs, sizeof(TotalNPCs)));
-		for (int i = 0; i < TotalNPCs; i++)
+		return;
+	}
+
+	// Sort the clusters by "distance from centre of area" ascending.
+	// This ensures we iterate from closest to furthest cluster.
+	ClusterPoints.Sort([this](const FVector& Point1, const FVector& Point2)
+	{
+		const float Distance1 = FVector::Distance(Point1, WorldPosition);
```

Comment:
uber unnecessary performance nit here - but you can use DistSquared to prevent the sqrt cost.